modules/pam_filter/pam_filter.c

```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
 	}
 
 	strcpy(levp[0], ARGS_NAME);
-	size = ARGS_OFFSET;
 	for (i=0; i<argc; ++i) {
 	    if (i)
-		levp[0][size++] = ' ';
-	    strcpy(levp[0]+size, argv[i]);
-	    size += strlen(argv[i]);
+		strcat(levp[0], " ");
+	    strcat(levp[0], argv[i]);
 	}
 
 	/* the "SERVICE" variable */
 
-#define SERVICE_NAME      "SERVICE="
-#define SERVICE_OFFSET    (sizeof(SERVICE_NAME) - 1)
-
 	retval = pam_get_item(pamh, PAM_SERVICE, &tmp);
 	if (retval != PAM_SUCCESS || tmp == NULL) {
 	    pam_syslog(pamh, LOG_CRIT, "service name not found");
-	    if (levp) {
-		free(levp[0]);
-		free(levp);
-	    }
+	    free(levp[0]);
```

Comment:
This still needs the `if (levp)` check.

Reply:
Impossible for levp to be NULL here. See lines above: https://github.com/linux-pam/linux-pam/blob/b07b6374bc1a5ce11d1a767f7d5cbcc53bd15480/modules/pam_filter/pam_filter.c#L123